pubsub/pubsub_goredis.go

```diff
@@ -83,88 +85,58 @@ func (ps *GoRedisPubSub) Stop() error {
 	return nil
 }
 
-// assert that GoRedisPubSub implements PubSub
-var _ PubSub = (*GoRedisPubSub)(nil)
-
-// when a topic is created, it is stored in the topics map and a goroutine is
-// started to listen for messages on the topic; each message is sent to all
-// subscribers to the topic.
-func (ps *GoRedisPubSub) NewTopic(ctx context.Context, topic string) Topic {
-	t := &GoRedisTopic{
-		client:      ps.client,
-		topic:       topic,
-		redisSub:    ps.client.Subscribe(ctx, topic),
-		subscribers: make([]chan string, 0),
-		done:        make(chan struct{}),
+func (ps *GoRedisPubSub) Close() {
+	ps.mut.Lock()
+	defer ps.mut.Unlock()
+	for _, sub := range ps.subs {
+		sub.Close()
 	}
+	ps.subs = nil
+	ps.client.Close()
+}
+
+func (ps *GoRedisPubSub) Publish(ctx context.Context, topic, message string) error {
+	ps.mut.RLock()
+	defer ps.mut.RUnlock()
+	return ps.client.Publish(ctx, topic, message).Err()
+}
 
+func (ps *GoRedisPubSub) Subscribe(ctx context.Context, topic string) Subscription {
+	ps.mut.Lock()
+	defer ps.mut.Unlock()
+	sub := &GoRedisSubscription{
+		topic:  topic,
+		pubsub: ps.client.Subscribe(ctx, topic),
+		ch:     make(chan string, 100),
+		done:   make(chan struct{}),
+	}
+	ps.subs = append(ps.subs, sub)
 	go func() {
+		redisch := sub.pubsub.Channel()
 		for {
 			select {
-			case msg := <-t.redisSub.Channel():
-				t.mut.RLock()
-				subscribers := t.subscribers
-				t.mut.RUnlock()
-				for _, sub := range subscribers {
-					select {
-					case sub <- msg.Payload:
-					case <-t.done:
-						return
-					}
+			case <-sub.done:
+				close(sub.ch)
+				return
+			case msg := <-redisch:
+				if msg == nil {
+					continue
 				}
-			case <-t.done:
-				t.redisSub.Close()
-				t.mut.RLock()
-				subscribers := t.subscribers
-				for _, sub := range subscribers {
-					close(sub)
+				select {
+				case sub.ch <- msg.Payload:
+				default:
```

Comment:
Can we log something here indicating a message has been dropped?

Reply:
I really didn't want to do that for a number of reasons, but also it's probably the cheapest way to detect a problem here, so I did it anyway.